nova/virt/vmwareapi/cluster_util.py

```diff
@@ -61,12 +58,39 @@ def _create_host_group_spec(client_factory, name, host_refs, operation="add",
     else:
         group.host = host_refs
 
+    return group
+
+
+def create_group_spec(client_factory, group, operation):
+    """Create a ClusterGroupSpec object"""
+    if operation not in ('add', 'edit', 'remove'):
+        msg = 'Invalid operation for ClusterGroupSpec: {}'.format(operation)
+        raise exception.ValidationError(msg)
+
     group_spec = client_factory.create('ns0:ClusterGroupSpec')
     group_spec.operation = operation
     group_spec.info = group
+    if operation == 'remove':
+        group_spec.removeKey = group.name
+
     return group_spec
 
 
+def _create_vm_group_spec(client_factory, group_info, vm_refs,
+                          operation="add", group=None):
+    if group:
+        # On vCenter UI, it is not possible to create VM group without
+        # VMs attached to it. But, using APIs, it is possible to create
+        # VM group without VMs attached. Therefore, check for existence
+        # of vm attribute in the group to avoid exceptions
+        if hasattr(group, 'vm'):
+            vm_refs += group.vm
```

Comment:
I would rather see here `vm_refs = vm_refs + group.vm`, as the current code modifies the passed `vm_refs`, which seems to me an unintended side-effect for the caller.

Reply:
This is the extracted code. There is 1 caller inside the module ;)

I agree with the "we should not change the list we get".